samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java

```diff
@@ -59,30 +63,34 @@ public void onNext(M message, MessageCollector collector, TaskCoordinator coordi
     K key = thisPartialJoinFn.getKey(message);
     thisPartialJoinFn.getState().put(key, new PartialJoinMessage<>(message, System.currentTimeMillis()));
     PartialJoinMessage<JM> otherMessage = otherPartialJoinFn.getState().get(key);
-    if (otherMessage != null) {
+    long now = System.currentTimeMillis();
+    if (otherMessage != null && otherMessage.getReceivedAt() > now - ttlMs) {
       RM joinResult = thisPartialJoinFn.apply(message, otherMessage.getMessage());
       this.propagateResult(joinResult, collector, coordinator);
     }
   }
 
   @Override
   public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    long now = System.currentTimeMillis();
+
     KeyValueStore<K, PartialJoinMessage<M>> thisState = thisPartialJoinFn.getState();
     KeyValueIterator<K, PartialJoinMessage<M>> iterator = thisState.all();
     List<K> keysToRemove = new ArrayList<>();
-    long now = System.currentTimeMillis();
 
     while (iterator.hasNext()) {
       Entry<K, PartialJoinMessage<M>> entry = iterator.next();
-      if (entry.getValue().getTimestamp() < now - ttlMs) {
+      if (entry.getValue().getReceivedAt() < now - ttlMs) {
         keysToRemove.add(entry.getKey());
       } else {
         break;
       }
     }
 
+    iterator.close();
     thisState.deleteAll(keysToRemove);
 
+    LOGGER.debug("onTimer self time: {} ms", System.currentTimeMillis() - now);
```

Comment:
How often do we trigger onTimer()? I don't think that it would be often, right? I would suggest keeping this at info level until we have metrics reporting. One more suggestion: let's print the operator name and ID here as well.